src/tfidf.rs

```diff
@@ -0,0 +1,172 @@
+use super::Jieba;
+use hashbrown::HashMap;
+use lazy_static::lazy_static;
+use std::collections::{BTreeSet, BinaryHeap};
+use std::io::{self, BufRead, BufReader};
+
+lazy_static! {
+    static ref STOP_WORDS: BTreeSet<String> = {
+        let mut set = BTreeSet::new();
+        let words = [
+            "the", "of", "is", "and", "to", "in", "that", "we", "for", "an", "are", "by", "be", "as", "on", "with",
+            "can", "if", "from", "which", "you", "it", "this", "then", "at", "have", "all", "not", "one", "has", "or",
+            "that",
+        ];
+
+        for &s in words.iter() {
+            set.insert(String::from(s));
+        }
+
+        set
+    };
+}
+
+static DEFAULT_IDF: &str = include_str!("data/idf.txt");
+
+pub trait KeywordExtract {
+    fn extract_tags<'a>(&'a self, _: &'a str, _: usize, _: Vec<String>) -> Vec<String>;
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord)]
+struct HeapNode<'a> {
+    tfidf: u64, //using u64 but not f64 so that it conforms to Ord
+    word: &'a str,
+}
+
+#[derive(Debug)]
+pub struct TFIDF {
+    jieba: Jieba,
+    idf_dict: HashMap<String, u64>,
+    median_idf: u64,
+}
+
+impl TFIDF {
+    pub fn new() -> Self {
+        let mut instance = TFIDF {
```

Comment:
If we are gonna expose it through `Jieba::tfidf`, I'd make `jieba` as reference:

```rust
pub struct TFIDF<'a> {
    jieba: &'a Jieba,
    idf_dict: HashMap<String, u64>,
    median_idf: u64,
}

impl TFIDF {
    pub fn new(jieba: &Jieba) ->Self {
          let mut instance = Self {
              jieba,
              idf_dict: HashMap::new(),
              median_idf: 0
          };
          // load dict...
          instance
    }
}
```

Reply:
I'll add `new_with_jieba(jieba: Jieba)` to handle the case where you already have a constructed instance and wouldn't like to construct and load the dictionary again, and have `new` being called it at the lazy_static, so that we don't have to struggle with having another static instance of Jieba to pass it into the static instance of TFIDF.

Reply:
The problem with this approach is that `new_with_jieba(jieba: Jieba)` take the ownership of the constructed instance and move it to `TFIDF`, after that user can not access the existing `Jieba` instance.

Reply:
I've changed it to `new_with_jieba(jieba: &Jieba)` and make lazy_static pass another static instance of Jieba. However, in order to support `jieba.enable_parallel` as the python version did. I still feel that it would be better to make a copy of it, and make the dict HashMap protected by a RwLock and shared among tasks (suppose we are implementing the enable_parallel by workers style). I felt like `struct Jieba` should hold the states for induction so it should be shared-nothing among TFIDF tasks but only shared the dict. Though right now `Jieba` doesn't hold any data beside `total` and `dict`.

BTW, if we are following Python's version, `Jieba` should be renamed as `Tokenizer`. `Jieba` is actually a proxy to a module instance of Tokenizer.

Reply:
We don't want to follow Python's version. I don't like the `DEFAULT_JIEBA` instance because it's not muable, if user want to add customized words they have to create their own instance, it does not work like the Python's version.

It's totally fine to ask the user to provide `Jieba` instance when creating `TFIDF` instance.

Reply:
`DEFAULT_JIEBA` is needed because `fn tfidf` needs a static instance of TFIDF, which is `DEFAULT_TFIDF_KEYWORD_EXTRACTOR`. In that case, I should remove the `fn tfidf` and make user construct the `TFIDF` and use `extract_tags` method. Does that sound good to you?  (`fn tfidf` was there because python version has it, it exposes extract_tags from tfidf to as part of jieba)

Reply:
Sounds good. Since this is written in Rust, we don't need to exactly mimic the behaviors of the Python version.

Reply:
Awesome, let me change the code and push it again.